src/core/parser.js

```diff
@@ -532,25 +512,19 @@ class Parser {
       default:
         length = this.findDefaultInlineStreamEnd(stream);
     }
-    let imageStream = stream.makeSubStream(startPos, length, dict);
 
     // Cache all images below the MAX_LENGTH_TO_CACHE threshold by their
-    // adler32 checksum.
+    // strinified content, to prevent possible hash collisions.
     let cacheKey;
-    if (length < MAX_LENGTH_TO_CACHE && dictLength < MAX_ADLER32_LENGTH) {
-      const imageBytes = imageStream.getBytes();
-      imageStream.reset();
-
+    if (length < MAX_LENGTH_TO_CACHE && dictLength > 0) {
       const initialStreamPos = stream.pos;
       // Set the stream position to the beginning of the dictionary data...
       stream.pos = lexer.beginInlineImagePos;
-      // ... and fetch the bytes of the *entire* dictionary.
-      const dictBytes = stream.getBytes(dictLength);
+      // ... and fetch the byte-string of the dictionary *and* the inline image.
+      cacheKey = stream.getString(dictLength + length);
```

Comment:
I suppose that if the stream has a length of N then the string will have a length equal to N.
But if I'm correct, js strings are internally saved in UTF-16 which means that it will take 2 * N bytes.
So if possible we should get char codes from uint16 values and then generate a `ceil(N/2)`-length string.

Reply:
Just something like:
https://github.com/mozilla/pdf.js/blob/master/src/core/base_stream.js#L83
but instead of using:
https://github.com/mozilla/pdf.js/blob/f7449563ef8722df30c2988fe571236ab9eb89f9/src/shared/util.js#L589
use something like:
```js
for (let i = 0; .i < ii; i += 2) {
  strBuf.push(String.fromCharCode(buf[i]<< 8 | buf[i + 1]));
}
```
If data are correctly aligned, it should be possible to create a `Uin16Array` view from the `Uint8Array` and the just iterate over it to get the char code: this way the loop size will be divide by 2.

Reply:
That's certainly clever!

However, what happens when there's an odd number of elements in the returned bytes?
Testing this algorithm, it seems to me that the following (made up) examples would both generate *the same* string despite the data definitely not being idential:
 - `new Uint8Array([1, 2, 3, 4, 5, 0])`
 - `new Uint8Array([1, 2, 3, 4, 5]`

A possibly very simple way of addressing that could perhaps be to add the original `length` to the cacheKey?

Reply:
Yep the length should be fine. Maybe `buf[i + 1]<< 8 | buf[i])` and just take `buf[i]` if `i == length - 1` could be fine too.

Reply:
Thank you!

Reply:
After thinking about that, I think we should add the length to the key.